### DISCUSSION TITLE
Basically, doesn't it support running through mono in net472 version?

### DISCUSSION THREAD

**Discussion by kdaek21** [ikvmnet/ikvm#615] [Q&A]
First of all, since Linux does not have net framework, when you unzip the net472 Linux version, there are only libXXX.so files in the bin folder.

Before .net core came out, the mono project provided mono to run .net framework assemblies on non-Windows systems. (It seems that support has been discontinued now that .net core has been introduced, but the last version can still be downloaded and installed/used.)
So, you can run .net assemblies through mono as follows:

'mono xxx.exe'

To check if IKVM also runs on mono in this way, I ran it in the format of
'mono [ikvm-jre-path]/bin/java.exe [Java class with main function]'

or
'mono [ikvm-jre-path]/bin/java.exe -jar [jar file]'

Then, in the default state, it raises an exception while searching for libraries in runtimes, as in the attached txt files.

Windows : 
[Win-Mono-IKVM-8.10.3-jre-net472-win-x64-default-running-Exception.txt](https://github.com/user-attachments/files/17615090/Win-Mono-IKVM-8.10.3-jre-net472-win-x64-default-running-Exception.txt)
LInux : 
[Linux-Mono-IKVM-8.10.3-jre-net472-win-x64-default-running-Exception.txt](https://github.com/user-attachments/files/17615094/Linux-Mono-IKVM-8.10.3-jre-net472-win-x64-default-running-Exception.txt)

So I unzipped IKVM.8.10.X.nupkg, extracted the runtimes, and ran it. In special cases like Spring Boot, an exception occurs as in the attached file (it seems to be a problem with mono itself), but general Java class files and executable jar files run normally.
[Win-Mono-IKVM-8.10.3-jre-net472-win-x64-runtimesAdded-SpringBoot-running-Exception.txt](https://github.com/user-attachments/files/17615121/Win-Mono-IKVM-8.10.3-jre-net472-win-x64-runtimesAdded-SpringBoot-running-Exception.txt)

The Windows version is supposed to look for the Windows file system, but it doesn't work on Linux, throwing an exception like the one in the attached file.
[LInux-Mono-IKVM-8.10.3-jre-net472-win-x64-runtimesAdded-running-Exception.txt](https://github.com/user-attachments/files/17615155/LInux-Mono-IKVM-8.10.3-jre-net472-win-x64-runtimesAdded-running-Exception.txt)

Is it not possible to make it work without the runtimes folder through mono? (ikvm.dll in runtimes\win-x64\native is also included in the bin folder by default for Windows.)

Also, can't the net472 version be used in a dynamic conversion method using java.exe through mono on Linux?

* The test on Linux combined 3 packages.
(IKVM-8.10.3-jre-net472-win-x64 + IKVM.8.10.3's runtimes + libXXX.so in IKVM-8.10.3-jre-net472-linux-x64's bin)

**Comment by wasabii** [comment]
Title is correct. I no longer distribute a version that would work on Framework Mono.

The only part that does work and that I do test is ikvmc itself, as this is required for cross compiling scenarios.

**Reply by wasabii** [reply]
FYI, I would not mind fixing this. IKVM obviously used to work on Mono. But there's so many changes over the last few years. And using it on Mono has been such a low demand thing since Core, that I decide not to hold up any releases over Mono support, or to add automated testing for Mono, etc. 

Is be happy if somebody wanted to do the work of getting it going again

**Reply by kdaek21** [reply]
Even the Mono site itself gradually migrated to .NET Core and said that they would no longer provide additional support once the migration was complete, so they strongly advised that we migrate to .NET Core. However, after doing some simple tests, I found that assemblies made based on .NET Framework and .NET Core have different structures, and assemblies made based on .NET Framework cannot be run with the dotnet command. (It feels like they only support build.) So, if you want to use a program made based on .NET Framework on a non-Windows system, you still need mono. 

If you want to use IKVM like general Java, you can use IKVM-jdk or IKVM-jre by replacing it with a version higher than net6.0. In the case of static conversion, since you usually have the source of the Wrapper part, you can use it by reconverting it. So in general cases, there is no longer a need to support operation through mono. However, there may be people like me who insist on running the net472 version through mono, or there may be people who want to use the converted net472 version on Linux through mono without reconverting it to a version higher than net6.0. So I hope that someone will create something that works on mono again.

**Reply by kdaek21** [reply]
FYI, I thought that .NET assemblies compiled for net472 or lower would not run in the .NET Core environment, but after looking into it more and testing, I found that if you create [executable file name].runtimeconfig.json to match the currently installed dotnet core runtime, you can run it in the format of 'dotnet [executable file name].exe' in the dotnet core environment. (It's not that the assembly structure is different, but in order to run it through dotnet core, [executable file name].runtimeconfig.json file containing target framework settings and other necessary settings is required.)

So I tried to run it for net472 in the dotnet core environment, so I created java.runtimeconfig.json and tried dotnet java.exe, but an error related to System.Text.Json occurred, so I looked it up and just in case, I replaced IKVM.Runtime with IKVM.Runtime in net8.0 and tried to run it, and it worked normally. So this time, I tried to retest the net472 version for Windows on Linux through mono, which failed last time, through the dotnet core environment, so I added the native files libXXXX.so from the Linux version and tried 'dotnet java.exe' in the Linux environment. When I tried to run it on Linux through mono last time, an error related to WinNTFileSystem.java that occurred again popped up, so when I analyzed the rt.jar in the original Java 8, some classes Since the configuration is different depending on the OS environment you are using, and IKVM.Java, which plays the role of rt.jar in IKVM, follows it, it is impossible to run Windows on Linux via mono or dotnet core. Therefore, I knew that it was absolutely impossible to run something compiled for net472 in a non-Windows environment via mono (runtimes folder required and there was no IKVM.Java for net472 compiled for non-Windows). So I found out that if you replace IKVM.Runtime and IKVM.Java and native libraries according to the OS and dotnet core environment you want to use, and appropriately create [executable file name].runtimeconfig.json, you can use it in the dotnet core environment.
